python/ray/air/execution/resources/resource_manager.py

```diff
@@ -0,0 +1,106 @@
+import abc
+
+from typing import List, Optional
+
+import ray
+from ray.air.execution.resources.request import (
+    ResourceRequest,
+    AllocatedResource,
+)
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class ResourceManager(abc.ABC):
+    """Resource manager interface.
+
+    A resource manager can be used to request resources from a Ray cluster and
+    allocate them to remote Ray tasks.
+
+    Because Ray clusters can autoscale, we need to request resources before
+    we can allocate them. Effectively this means we have to keep track
+    of requested resources (which can trigger autoscaling). Once the requested
+    resources are available to be scheduled, the control loop can acquire them.
```

Comment:
Same question about resource state - we should make it extremely nonambiguous what requested/allocated/available/etc. mean.

One random idea is to remove `AllocatedResource` altogether and have a `state` attribute of `ResourceRequest`, which has a predefined list of state enum values.

`annotate_remote_objects` would validate that `state == AVAILABLE` (or whatever the exact state is to indicate that it is requested but not allocated yet).

Reply:
We should keep requests separate from allocated resources. 

To think of it in pseudo C terms:

```
// Resource requests
int rq1 = 5
int rq2 = 5

assert(rq1 == rq2)

// Allocated memory
ar = malloc(rq1)

// ... use memory allocated in ar somewhere

free(ar)
```

For a more high-level explanation:

If resource state is tied to the request (e.g. via a property), we lose the equality of two different resource objects with the same specs. This in turn would mean that we lose the flexible association of resources to actors. And this in turn would mean that Tune can't control which trials it wants to run first.

Another reason is that it would change the ownership contract. A ResourceRequest is a declarative statement. It is associated to trials (as a PGF), it can be shipped around the cluster, copied, is part of several APIs. The allocated resource however is a singleton. Using a resource manager at all means.

I'll expand the resource manager docstring a bit

Reply:
Ah I see, that makes a lot of sense why we need a wrapper object!

I was definitely also confused by the fact that we only created a new object once the resource is allocated/acquired. 

To get a better understanding of it, I tried to extend it for the ResourceManager to be able to track the requests throughout their entire lifecycle - I think the end result would help make the state machine more clear, but it would also make the bookkeeping logic more complex. Just leaving it here for reference... 😄 
```python
class ResourceRequestState(Enum):
    REQUESTED = auto() # Requested by the ResourceManager.
    READY = auto() # Ready and available for scheduling.
    ACQUIRED = auto() # Acquired by a task/actor.

@dataclass
class ManagedResourceRequest:
    request: ResourceRequest
    state: ResourceRequestState = ResourceRequestState.REQUESTED
    associated_objects: List[RemoteRayObject] = []  # not sure if this is needed

    def mark_ready(self):
        self.state = ResourceRequestState.READY

    def annotate_remote_objects(self, objects: List[RemoteRayObject]) -> List[RemoteRayObject]:
        raise NotImplementedError

    def mark_acquired(self, objects: List[RemoteRayObject]):
        annotated_objects = self.annotate_remote_objects(objects)
        self.associated_objects = annotated_objects
        self.state = ResourceRequestState.ACQUIRED
        return annotated_objects
```